File: examples/utils/setupFonts.c
#include <float.h>
#include <stdio.h>
#include <stdlib.h>
#include <strings.h>

#define CIMGUI_DEFINE_ENUMS_AND_STRUCTS
#include "cimgui.h"
#include "utils.h"
#include "setupFonts.h"

#define ImGui_GetIO igGetIO

#define MAX_PATH  2048
const char* IconFontPath = "../utils/fonticon/fa6/fa-solid-900.ttf";
char sBufFontPath[MAX_PATH];

char WinFontNameTbl[][MAX_PATH] = {
                                  "meiryo.ttc"     // Windows 7,8
                                 ,"YuGothM.ttc"    // Windows 10
                                 ,"segoeui.ttf"    // English standard
                                 };
char LinuxFontNameTbl[][MAX_PATH] = { // For Linux Mint 22 (Ubuntu/Debian family ok ?)
                            "/usr/share/fonts/opentype/noto/NotoSansCJK-Regular.ttc"          // JP
                           ,"/usr/share/fonts/opentype/ipafont-gothic/ipag.ttf"               // Debian
                           ,"/usr/share/fonts/opentype/ipafont-gothic/ipam.ttf"               // Debian
                           ,"/usr/share/fonts/truetype/liberation/LiberationMono-Regular.ttf" // Linux Mint English
                           };

/*-----------------
 * getWinFontPath()
 *----------------*/
char* getWinFontPath(char* sBuf, int bufSize, const char* fontName) {
  char* sWinDir = getenv("windir");
  if (sWinDir == NULL) return NULL;
  snprintf(sBuf, bufSize, "%s\\Fonts\\%s", sWinDir, fontName);
  return sBuf;
}

/*------------
 * point2px()
 *-----------*/
float point2px(float point) { //## Convert point to pixel
  return (point * 96) / 72;
}

const ImWchar ranges_icon_fonts[]  = {(ImWchar)ICON_MIN_FA, (ImWchar)ICON_MAX_FA, (ImWchar)0};
/*--------------
 * setupFonts()
 *-------------*/
void setupFonts(void) {
  ImGuiIO* pio = ImGui_GetIO();
  ImFontConfig* config  = ImFontConfig_ImFontConfig();
  ImFont* font = NULL;
  char* fontPath;
  int tableLen = sizeof(WinFontNameTbl) / MAX_PATH;
  for(int i=0; i<tableLen; i++){
    fontPath = getWinFontPath(sBufFontPath, sizeof(sBufFontPath), WinFontNameTbl[i]);
    if (existsFile(fontPath)) {
      font = ImFontAtlas_AddFontFromFileTTF(pio->Fonts, fontPath, point2px(14.5)
          , config
          , ImFontAtlas_GetGlyphRangesJapanese(pio->Fonts));
      printf("Found FontPath: [%s]\n",fontPath);
      break;
    }
  }
  tableLen = sizeof(LinuxFontNameTbl) / MAX_PATH;
  for(int i=0; i<tableLen; i++){
    fontPath = LinuxFontNameTbl[i];
    if (existsFile(fontPath)) {
      font = ImFontAtlas_AddFontFromFileTTF(pio->Fonts, fontPath, point2px(13)
          , config
          , ImFontAtlas_GetGlyphRangesJapanese(pio->Fonts));
      printf("Found FontPath: [%s]\n",fontPath);
      break;
    }
  }
  if (font == NULL) {
    printf("Error!: Font loading falied: in %s\n", __FILE__);
    printf("Default has been set.\n");
    ImFontAtlas_AddFontDefault(pio->Fonts, NULL);
  }
  // Merge IconFont
  config->MergeMode = true;
  ImFontAtlas_AddFontFromFileTTF(pio->Fonts, IconFontPath, point2px(11), config , ranges_icon_fonts);
}

File: src/libzig/setupfont/src/setupFonts.zig
const std = @import("std");
const ig = @import("cimgui");

const MAX_PATH = 2048;
const IconFontPath = "resources/fonticon/fa6/fa-solid-900.ttf";

var sBufFontPath: [MAX_PATH]u8 = undefined;

const WinFontNameTbl = [_][]const u8{
    "meiryo.ttc",    // Windows 7,8
    "YuGothM.ttc",   // Windows 10
    "segoeui.ttf",   // English standard
};

const LinuxFontNameTbl = [_][]const u8{
    "/usr/share/fonts/opentype/ipafont-gothic/ipag.ttf",               // Debian jp
    "/usr/share/fonts/opentype/noto/NotoSansCJK-Regular.ttc",          // JP
    "/usr/share/fonts/opentype/ipafont-gothic/ipam.ttf",               // Debian jp
    "/usr/share/fonts/truetype/liberation/LiberationMono-Regular.ttf", // Linux Mint English
    "/usr/share/fonts/truetype/dejavu/DejaVuSansMono.ttf",             // English region standard font
};

/// Check if file exists
fn existsFile(path: []const u8) bool {
    const file = std.fs.cwd().openFile(path, .{}) catch return false;
    file.close();
    return true;
}

/// Get Windows font path
fn getWinFontPath(buf: []u8, fontName: []const u8) ?[]const u8 {
    const windir = std.process.getEnvVarOwned(std.heap.page_allocator, "windir") catch return null;
    defer std.heap.page_allocator.free(windir);

    const result = std.fmt.bufPrint(buf, "{s}\\Fonts\\{s}", .{ windir, fontName }) catch return null;
    return result;
}

/// Convert point to pixel
fn point2px(point: f32) f32 {
    return (point * 96.0) / 72.0;
}

const ICON_MIN_FA: c_int = 0xf000;
const ICON_MAX_FA: c_int = 0xf8ff;

const ranges_icon_fonts = [_]ig.ImWchar{ ICON_MIN_FA, ICON_MAX_FA, 0 };

/// Setup fonts for ImGui
pub fn setupFonts() ?*ig.ImFont {
    const pio = ig.igGetIO_Nil();
    const config = ig.ImFontConfig_ImFontConfig();
    var font: ?*ig.ImFont = null;

    // Try Windows fonts
    for (WinFontNameTbl) |fontName| {
        if (getWinFontPath(&sBufFontPath, fontName)) |fontPath| {
            if (existsFile(fontPath)) {
                font = ig.ImFontAtlas_AddFontFromFileTTF( pio.*.Fonts, fontPath.ptr, point2px(14.5), null, null,);
                std.debug.print("Found FontPath: [{s}]\n", .{fontPath});
                break;
            }
        }
    }

    // If not found, try Linux fonts
    if (font == null) {
        for (LinuxFontNameTbl) |fontPath| {
            if (existsFile(fontPath)) {
                font = ig.ImFontAtlas_AddFontFromFileTTF( pio.*.Fonts, fontPath.ptr, point2px(13.0), null, null,);
                std.debug.print("Found FontPath: [{s}]\n", .{fontPath});
                break;
            }
        }
    }

    // If still not found, use default
    if (font == null) {
        std.debug.print("Error!: Font loading failed\n", .{});
        std.debug.print("Default has been set.\n", .{});
        _ = ig.ImFontAtlas_AddFontDefault(pio.*.Fonts, null);
    }

    // Merge IconFont
    config.*.MergeMode = true;
    return ig.ImFontAtlas_AddFontFromFileTTF( pio.*.Fonts, IconFontPath, point2px(11.0), config, &ranges_icon_fonts,);
}
